util/network/cniprovider/cni.go

```diff
@@ -67,19 +79,114 @@ func (c *cniProvider) initNetwork() error {
 		}
 		defer l.Unlock()
 	}
-	ns, err := c.New("")
+	ns, err := c.New(context.TODO(), "")
 	if err != nil {
 		return err
 	}
 	return ns.Close()
 }
 
-func (c *cniProvider) New(hostname string) (network.Namespace, error) {
+type cniPool struct {
+	provider   *cniProvider
+	mu         sync.Mutex
+	targetSize int
+	actualSize int
+	available  []*cniNS
+}
+
+func (pool *cniPool) fillPool(ctx context.Context) {
+	for {
+		pool.mu.Lock()
+		actualSize := pool.actualSize
+		pool.mu.Unlock()
+		if actualSize >= pool.targetSize {
+			return
+		}
+		if _, err := pool.getNew(ctx); err != nil {
+			bklog.G(ctx).Errorf("failed to create new network namespace while prefilling pool: %s", err)
+			return
+		}
+	}
+}
+
+func (pool *cniPool) get(ctx context.Context) (*cniNS, error) {
+	pool.mu.Lock()
+	if len(pool.available) > 0 {
+		ns := pool.available[0]
+		pool.available = pool.available[1:]
+		pool.mu.Unlock()
+		trace.SpanFromContext(ctx).AddEvent("returning network namespace from pool")
+		bklog.G(ctx).Debugf("returning network namespace %s from pool", ns.id)
+		return ns, nil
+	}
+	pool.mu.Unlock()
+
+	return pool.getNew(ctx)
+}
+
+func (pool *cniPool) getNew(ctx context.Context) (*cniNS, error) {
+	ns, err := pool.provider.newNS(ctx, "")
+	if err != nil {
+		return nil, err
+	}
+	ns.pool = pool
+
+	pool.mu.Lock()
+	pool.actualSize++
+	pool.mu.Unlock()
+	return ns, nil
+}
+
+func (pool *cniPool) put(ns *cniNS) {
+	putTime := time.Now()
+	ns.lastUsed = putTime
+
+	pool.mu.Lock()
+	pool.available = append(pool.available, ns)
+	actualSize := pool.actualSize
+	pool.mu.Unlock()
+
+	if actualSize > pool.targetSize {
+		// We have more network namespaces than our target number, so
+		// release this extra one after some time if it hasn't been
+		// reused by then.
+		time.AfterFunc(aboveTargetGracePeriod, func() {
+			pool.mu.Lock()
+			if pool.actualSize > pool.targetSize && ns.lastUsed.Equal(putTime) {
+				for i, poolNS := range pool.available {
+					if poolNS == ns {
```

Comment:
I'm not sure if this is quite correct. You do `actualSize > pool.targetSize` comparison when deciding to add a cleanup func and also inside the cleanup but I don't think there is a guarantee that these conditions are the same when these places get called for a specific ns.

At least it seems safer if we skip this check end just always call `cleanUpToDefaultSize()` that is a no-op if available is not greater than size.

Reply:
The namespaces are reused in FIFO order, so I think this is correct as written. If we were not above the target size at `put` time, it will be some other namespace that brings us the target size, and that's the one we'll want to release if it hasn't been reused after 5 minutes.

But thinking about this some more, I don't think we want the FIFO order. It means we'll cycle through the namespaces and it will be hard to shrink back to the target size if there's constant activity. LIFO seems better. So I've rewritten this to use LIFO and align more closely to your suggestion.